### PR TITLE
CGAL: more viewers

### PR DESCRIPTION
**Pull request by gdamiand**
_Please use the following template to help us managing pull requests._

## Summary of Changes

Improve basic viewers:
* for 2D objects, fix the viewer in 2D
* draw method are now specialized by the data structure, allowing to use different draw in a same program

No change in api.

Add 3 new viewers:
* point set
* polygon
* polygon with hole

## Release Management

* Affected package(s): 3D Point set, 2D polygon
* Feature/Small Feature (if any): [here](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/More_basic_viewers) - **pre-approval date:** 2019/06/25
* Link to compiled documentation (obligatory for small feature) [More viewers](https://perso.liris.cnrs.fr/guillaume.damiand/CGAL/More-viewers/Manual/)
* License and copyright ownership:

### CODE REVIEW

File: Point_set_3/doc/Point_set_3/CGAL/draw_point_set_3.h [hunk 1]
```diff
@@ -0,0 +1,14 @@
+namespace CGAL {
+  
+/*!
+\ingroup PkgDrawPointSet3D
+
+Open a new window and draw `aps`, an instance of the `CGAL::Point_set_3` class. The function is blocking, that is the program continues as soon as the user closes the window. This function requires CGAL_Qt5, and is only available if the flag CGAL_USE_BASIC_VIEWER is defined at compile time.
```
Comment (sloriot) [inline Point_set_3/doc/Point_set_3/CGAL/draw_point_set_3.h]:
```suggestion
opens a new window and draw `aps`, an instance of the `CGAL::Point_set_3` class. The function is blocking, that is the program continues as soon as the user closes the window. This function requires CGAL_Qt5, and is only available if the flag CGAL_USE_BASIC_VIEWER is defined at compile time.
```

File: Polygon/doc/Polygon/CGAL/draw_polygon_2.h [hunk 2]
```diff
@@ -0,0 +1,14 @@
+namespace CGAL {
+  
+/*!
+\ingroup PkgDrawPolygon2
+
+Open a new window and draw `ap`, an instance of the `CGAL::Polygon_2` class. The function is blocking, that is the program continues as soon as the user closes the window. This function requires CGAL_Qt5, and is only available if the flag CGAL_USE_BASIC_VIEWER is defined at compile time.
```
Comment (sloriot) [inline Polygon/doc/Polygon/CGAL/draw_polygon_2.h]:
```suggestion
opens a new window and draw `ap`, an instance of the `CGAL::Polygon_2` class. The function is blocking, that is the program continues as soon as the user closes the window. This function requires CGAL_Qt5, and is only available if the flag CGAL_USE_BASIC_VIEWER is defined at compile time.
```

File: Polygon/doc/Polygon/CGAL/draw_polygon_with_holes_2.h [hunk 3]
```diff
@@ -0,0 +1,14 @@
+namespace CGAL {
+  
+/*!
+\ingroup PkgDrawPolygonWithHoles2
+
+Open a new window and draw `aph`, an instance of the `CGAL::Polygon_with_holes_2` class. The function is blocking, that is the program continues as soon as the user closes the window. This function requires CGAL_Qt5, and is only available if the flag CGAL_USE_BASIC_VIEWER is defined at compile time.
```
Comment (sloriot) [inline Polygon/doc/Polygon/CGAL/draw_polygon_with_holes_2.h]:
```suggestion
opens a new window and draw `aph`, an instance of the `CGAL::Polygon_with_holes_2` class. The function is blocking, that is the program continues as soon as the user closes the window. This function requires CGAL_Qt5, and is only available if the flag CGAL_USE_BASIC_VIEWER is defined at compile time.
```

File: Linear_cell_complex/doc/Linear_cell_complex/Linear_cell_complex.txt [hunk 4]
```diff
@@ -253,7 +253,7 @@ We can observe that the second run is faster than the first one. Indeed, updatin
 \subsection Linear_cell_complexDraw Draw a Linear Cell Complex
 \anchor ssecDrawLCC
 
-A linear cell complex can be visualized by calling the `CGAL::draw()` function as shown in the following example. This function opens a new window showing the given linear cell complex. The function is blocking, that is the program continues as soon as the user closes the window.
+A linear cell complex can be visualized by calling the \link PkgDrawLinearCellComplex CGAL::draw<LCC>() \endlink function as shown in the following example. This function opens a new window showing the given linear cell complex. The function is blocking, that is the program continues as soon as the user closes the window.
```
Comment (sloriot) [inline Linear_cell_complex/doc/Linear_cell_complex/Linear_cell_complex.txt]:
```suggestion
A linear cell complex can be visualized by calling the \link PkgDrawLinearCellComplex CGAL::draw<LCC>() \endlink function as shown in the following example. This function opens a new window showing the given linear cell complex. A call to this function is blocking, that is the program continues as soon as the user closes the window.
```

File: Point_set_3/doc/Point_set_3/Point_set_3.txt [hunk 5]
```diff
@@ -163,6 +163,18 @@ overload provided for `CGAL::Point_set_3`:
 
 \cgalExample{Point_set_3/point_set_advanced.cpp}
 
+\subsection Point_set_3_Draw Draw a Point Set
+
+A 3D point set can be visualized by calling the \link PkgDrawPointSet3D CGAL::draw<PS>() \endlink function as shown in the following example. This function opens a new window showing the given point set. The function is blocking, that is the program continues as soon as the user closes the window.
```
Comment (sloriot) [inline Point_set_3/doc/Point_set_3/Point_set_3.txt]:
```suggestion
A 3D point set can be visualized by calling the \link PkgDrawPointSet3D CGAL::draw<PS>() \endlink function as shown in the following example. This function opens a new window showing the given point set. A call to this function is blocking, that is the program continues as soon as the user closes the window.
```

File: Polygon/doc/Polygon/CGAL/draw_polygon_2.h (hunk 2 above)
Comment (sloriot) [inline Polygon/doc/Polygon/CGAL/draw_polygon_2.h]:
```suggestion
Open a new window and draw `ap`, an instance of the `CGAL::Polygon_2` class. A call to this function is blocking, that is the program continues as soon as the user closes the window. This function requires CGAL_Qt5, and is only available if the flag CGAL_USE_BASIC_VIEWER is defined at compile time.
```

File: Polygon/doc/Polygon/CGAL/draw_polygon_with_holes_2.h (hunk 3 above)
Comment (sloriot) [inline Polygon/doc/Polygon/CGAL/draw_polygon_with_holes_2.h]:
```suggestion
Open a new window and draw `aph`, an instance of the `CGAL::Polygon_with_holes_2` class. A call to this function is blocking, that is the program continues as soon as the user closes the window. This function requires CGAL_Qt5, and is only available if the flag CGAL_USE_BASIC_VIEWER is defined at compile time.
```

File: Polygon/doc/Polygon/Polygon.txt [hunk 6]
```diff
@@ -70,6 +70,18 @@ and 3D Linear Geometric %Kernel.
 
 \cgalExample{Polygon/projected_polygon.cpp}
 
+\subsection subsecPolygonDraw Draw a Polygon
+
+A polygon can be visualized by calling the \link  PkgDrawPolygon2 CGAL::draw<P>() \endlink function as shown in the following example. This function opens a new window showing the given polygon. The function is blocking, that is the program continues as soon as the user closes the window (a version exists for polygon with holes, cf. \link PkgDrawPolygonWithHoles2 CGAL::draw<PH>() \endlink).
```
Comment (sloriot) [inline Polygon/doc/Polygon/Polygon.txt]:
```suggestion
A polygon can be visualized by calling the \link  PkgDrawPolygon2 CGAL::draw<P>() \endlink function as shown in the following example. This function opens a new window showing the given polygon. A call to this function is blocking, that is the program continues as soon as the user closes the window (a version exists for polygon with holes, cf. \link PkgDrawPolygonWithHoles2 CGAL::draw<PH>() \endlink).
```

File: Polyhedron/doc/Polyhedron/Polyhedron.txt [hunk 7]
```diff
@@ -278,7 +278,7 @@ are also marked in the program code.
 \subsection PolyhedronDraw Draw a Polyhedron
 \anchor ssecDrawPolyhedron
 
-A polyhedron can be visualized by calling the `CGAL::draw()` function as shown in the following example. This function opens a new window showing the given polyhedron. The function is blocking, that is the program continues as soon as the user closes the window.
+A polyhedron can be visualized by calling the \link PkgDrawPolyhedron CGAL::draw<POLY>() \endlink function as shown in the following example. This function opens a new window showing the given polyhedron. The function is blocking, that is the program continues as soon as the user closes the window.
```
Comment (sloriot) [inline Polyhedron/doc/Polyhedron/Polyhedron.txt]:
```suggestion
A polyhedron can be visualized by calling the \link PkgDrawPolyhedron CGAL::draw<POLY>() \endlink function as shown in the following example. This function opens a new window showing the given polyhedron. A call to this function is blocking, that is the program continues as soon as the user closes the window.
```

File: Surface_mesh/doc/Surface_mesh/Surface_mesh.txt [hunk 8]
```diff
@@ -347,7 +347,7 @@ refering to the right vertices.
 \section SurfaceMeshDraw Draw a Surface Mesh
 \anchor ssecDrawSurfaceMesh
 
-A surface mesh can be visualized by calling the `CGAL::draw()` function as shown in the following example. This function opens a new window showing the given surface mesh. The function is blocking, that is the program continues as soon as the user closes the window.
+A surface mesh can be visualized by calling the \link PkgDrawSurfaceMesh CGAL::draw<SM>() \endlink as shown in the following example. This function opens a new window showing the given surface mesh. The function is blocking, that is the program continues as soon as the user closes the window.
```
Comment (sloriot) [inline Surface_mesh/doc/Surface_mesh/Surface_mesh.txt]:
```suggestion
A surface mesh can be visualized by calling the \link PkgDrawSurfaceMesh CGAL::draw<SM>() \endlink as shown in the following example. This function opens a new window showing the given surface mesh. A call to this function is blocking, that is the program continues as soon as the user closes the window.
```

File: Triangulation_2/doc/Triangulation_2/Triangulation_2.txt [hunk 9]
```diff
@@ -475,7 +475,7 @@ Finally points on the convex hull are written to <TT>cout</TT>.
 \subsection Triangulation2Draw Draw a 2D Triangulation
 \anchor ssecDrawT2
 
-A 2D triangulation can be visualized by calling the `CGAL::draw()` function as shown in the following example. This function opens a new window showing the given 2D triangulation. The function is blocking, that is the program continues as soon as the user closes the window.
+A 2D triangulation can be visualized by calling the \link PkgDrawTriangulation2 CGAL::draw<T2>() \endlink function as shown in the following example. This function opens a new window showing the given 2D triangulation. The function is blocking, that is the program continues as soon as the user closes the window.
```
Comment (sloriot) [inline Triangulation_2/doc/Triangulation_2/Triangulation_2.txt]:
```suggestion
A 2D triangulation can be visualized by calling the \link PkgDrawTriangulation2 CGAL::draw<T2>() \endlink function as shown in the following example. This function opens a new window showing the given 2D triangulation. A call to this function is blocking, that is the program continues as soon as the user closes the window.
```

File: Triangulation_3/doc/Triangulation_3/Triangulation_3.txt [hunk 10]
```diff
@@ -538,7 +538,7 @@ removal of the first 100,000 vertices.
 \subsection Triangulation3Draw Draw a 3D Triangulation
 \anchor ssecDrawT3
 
-A 3D triangulation can be visualized by calling the `CGAL::draw()` function as shown in the following example. This function opens a new window showing the given 3D triangulation. The function is blocking, that is the program continues as soon as the user closes the window.
+A 3D triangulation can be visualized by calling the \link PkgDrawTriangulation3 CGAL::draw<T3>() \endlink function as shown in the following example. This function opens a new window showing the given 3D triangulation. The function is blocking, that is the program continues as soon as the user closes the window.
```
Comment (sloriot) [inline Triangulation_3/doc/Triangulation_3/Triangulation_3.txt]:
```suggestion
A 3D triangulation can be visualized by calling the \link PkgDrawTriangulation3 CGAL::draw<T3>() \endlink function as shown in the following example. This function opens a new window showing the given 3D triangulation. A call to this function is blocking, that is the program continues as soon as the user closes the window.
```

File: Linear_cell_complex/doc/Linear_cell_complex/Linear_cell_complex.txt (hunk 4 above)
Comment (sloriot) [inline Linear_cell_complex/doc/Linear_cell_complex/Linear_cell_complex.txt]:
```suggestion
A linear cell complex can be visualized by calling the \link PkgDrawLinearCellComplex CGAL::draw<LCC>() \endlink function as shown in the following example. This function opens a new window showing the given linear cell complex. A call to this function is blocking, that is the program continues as soon as the user closes the window.
```

**Comment by lrineau** [comment]
Small Travis error:
```
The following files do not have an include directive of the package license header:
Point_set_3/include/CGAL/draw_point_set_3.h
```
https://travis-ci.org/CGAL/cgal/jobs/547733417#L2775

**Comment by lrineau** [comment]
Small Travis error:
```
The following files do not have an include directive of the package license header:
Point_set_3/include/CGAL/draw_point_set_3.h
```
https://travis-ci.org/CGAL/cgal/jobs/547733417#L2775

**Comment by lrineau** [comment]
Small Travis error:
```
The following files do not have an include directive of the package license header:
Point_set_3/include/CGAL/draw_point_set_3.h
```
https://travis-ci.org/CGAL/cgal/jobs/547733417#L2775

**Comment by gdamiand** [comment]
Travis error during check package dependencies:
  Differences in Polygon: GraphicsView are new and not committed.
  Differences in Point_set_3: GraphicsView are new and not committed.
I don't know why I have these errors and how to solve them.
Any help please ?

**Comment by lrineau** [comment]
I have probably fixed that by commit 5538543faf4b2b39de9be1e44b9678a0949444eb I just pushed.

Now, `GraphicsView` is in the `dependencies` files for the following packages:
  - `GraphicsView`,
  - `Linear_cell_complex`,
  - `Point_set_3`,
  - `Polygon`,
  - `Polyhedron`,
  - `Surface_mesh`,
  - `Three`,
  - `Triangulation_2`, and
  - `Triangulation_3`.

Does that look good?

**Comment by gdamiand** [comment]
Thanks @lrineau ; I suppose it will fix the error. 
Let us wait the travis results.

**Comment by lrineau** [comment]
There are warnings and error, there:
https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-65.shtml#Surface_mesh_Examples

**Comment by maxGimeno** [comment]
I don't see how this branch broke links to Polygon from Minkowski_sum_2, but it seems it did :

```
/home/cgal-testsuite/cgal_doc_build/CGAL-5.0-Ic-68/doc/Minkowski_sum_2/Minkowski_sum_2.txt:209: warning: unable to resolve link to `Polygon_2' for \link command
/home/cgal-testsuite/cgal_doc_build/CGAL-5.0-Ic-68/doc/Minkowski_sum_2/Minkowski_sum_2.txt:214: warning: unable to resolve link to `Polygon_2' for \link command
/home/cgal-testsuite/cgal_doc_build/CGAL-5.0-Ic-68/doc/Minkowski_sum_2/Minkowski_sum_2.txt:215: warning: unable to resolve link to `Polygon_with_holes_2' for \link command
/home/cgal-testsuite/cgal_doc_build/CGAL-5.0-Ic-68/doc/Minkowski_sum_2/Minkowski_sum_2.txt:221: warning: unable to resolve link to `Polygon_with_holes_2' for \link command
```

**Comment by gdamiand** [comment]
@maxGimeno How do you conclude the problem comes from this branch ? The problem can come from another branch ?

I just ran a diff between this branch and master for both Polygon and Minkowski_sum_2 packages, and I found no reason for this error.

**Comment by lrineau** [comment]
I have bisected the error with the following script:
```shell
#! /bin/bash

rm -rf build-doc
mkdir build-doc
cd build-doc
cmake ../Documentation/doc
make -j10 doc
make Polygon_doc Minkowski_sum_2_doc |& (grep -F 'warning: unable to resolve link' || exit 0 && exit 1)
```
and the first bad commit is af32499a14958814a3c3e12c664c73b35915caf3.

**Comment by lrineau** [comment]
Now, here is the patch: https://github.com/CGAL/cgal/pull/4016/commits/29babbe4ad389b0e550adcad4e37d509cc904067.

The issue was the way default `INPUT` for Doxygen is computed by our CMake scripts: as soon as there exists a directory `doc/Polygon/CGAL/` (even if it is empty), then the directory `include/CGAL/` is ignored by Doxygen. And thus the addition of two files
```
Polygon/doc/Polygon/CGAL/draw_polygon_2.h
Polygon/doc/Polygon/CGAL/draw_polygon_with_holes_2.h
```
created the issue.

**Comment by gdamiand** [comment]
Thanks @lrineau ; there are some kind of magic in our documentation tools ;)

**Comment by sloriot** [comment]
@gdamiand why not documenting the functions in the source?

File: Polygon/doc/Polygon/Doxyfile.in [hunk 11]
```diff
@@ -3,7 +3,7 @@
 PROJECT_NAME = "CGAL ${CGAL_DOC_VERSION} - 2D Polygons"
 EXAMPLE_PATH = ${CGAL_PACKAGE_DIR}/examples \
                ${CGAL_Stream_support_EXAMPLE_DIR}
-
+INPUT += ${CGAL_PACKAGE_INCLUDE_DIR}/CGAL
```
Comment (lrineau) [inline Polygon/doc/Polygon/Doxyfile.in]: I broke the CGAL Doxygen testsuite with that patch. `${CGAL_PACKAGE_INCLUDE_DIR}/CGAL` evaluates to `...something.../include//CGAL`, and Doxygen-1.8.4 does not like the double `/`.

File: Polygon/doc/Polygon/Doxyfile.in [hunk 12]
```diff
@@ -3,7 +3,7 @@
 PROJECT_NAME = "CGAL ${CGAL_DOC_VERSION} - 2D Polygons"
 EXAMPLE_PATH = ${CGAL_PACKAGE_DIR}/examples \
                ${CGAL_Stream_support_EXAMPLE_DIR}
-INPUT += ${CGAL_PACKAGE_INCLUDE_DIR}/CGAL
+INPUT += ${CGAL_PACKAGE_INCLUDE_DIR}
```
Comment (lrineau) [inline Polygon/doc/Polygon/Doxyfile.in]: This patch is actually wrong. It works well in the Git-layout, but in a release, it picks the thousands of CGAL headers!

**Comment by sloriot** [comment]
Again, the doc should go in the source, no need to add more work for later.

**Comment by lrineau** [comment]
OK, let's postpone the testsuite of this PR until the doc is fixed.

**Comment by sloriot** [comment]
Should be good now.

**Comment by gdamiand** [comment]
Thanks @sloriot .

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-71.shtml

**Pull request merged**